src/com/sun/jna/internal/Cleaner.java

```diff
@@ -83,34 +59,43 @@ public synchronized Cleanable register(Object obj, Runnable cleanupTask) {
     }
 
     private synchronized CleanerRef add(CleanerRef ref) {
-        if(firstCleanable == null) {
-            firstCleanable = ref;
-        } else {
-            ref.setNext(firstCleanable);
-            firstCleanable.setPrevious(ref);
-            firstCleanable = ref;
+        synchronized (referenceQueue) {
+            if (firstCleanable == null) {
+                firstCleanable = ref;
+            } else {
+                ref.setNext(firstCleanable);
+                firstCleanable.setPrevious(ref);
+                firstCleanable = ref;
+            }
+            if (cleanerThread == null) {
```

Comment:
Does this need to be `volatile` to ensure it works correctly? Or is the presence of `start()` sufficient to ensure this is not a problem?

See any discussion of double checked locking for the relevant technical concern.

Reply:
I don't think `cleanerThread` needs to be volatile. I tried to get through the JLS regarding synchronization, but some google foo led me to this summary:

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility

> [...] [Chapter 17 of the Java Language Specification](https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5) defines the happens-before relation on memory operations such as reads and writes of shared variables. The results of a write by one thread are guaranteed to be visible to a read by another thread only if the write operation happens-before the read operation. The synchronized and volatile constructs, as well as the Thread.start() and Thread.join() methods, can form happens-before relationships. In particular:
>
> -   Each action in a thread happens-before every action in that thread that comes later in the program's order.
> -   An unlock (synchronized block or method exit) of a monitor happens-before every subsequent lock (synchronized block or method entry) of that same monitor. And because the happens-before relation is transitive, all actions of a thread prior to unlocking happen-before all actions subsequent to any thread locking that monitor.
 > [...]

In this case the critical monitor is the one associated with `referenceQueue`. Both places where the `cleanerThread` is modified are protected by a synchronization on `referenceQueue` and so to my reading of the happens-before guarantee for synchronized blocks, there can be no phantom-reads from `cleanerThread`. The same is true for `firstCleanable`, which needs to be considered also.

Usage of the `referenceQueue` monitor is sane from my POV, because there are two different variables affected (`cleanerThread` and `firstCleanable`) and multiple statement that must be executed atomically, which excludes the usage of a volatile variable.

Reply:
Yeah, I did enough rubber-ducking to convince myself the use of `referenceQueue` was rather effective.

My main concern was that the initial assignment of `cleanerThread` to an object happens-before, but there is no guarantee that the remainder of object initialization happens.... except in this case it must complete (within the block) in order for the `start()` method to be valid.